docs: add validate warnings and refactor the account's functions page (#1302)
starknet-io/starknet-docs: components/Starknet/modules/architecture-and-concepts/pages/accounts/account-functions.adoc

== Overview

The functions in the table xref:#starknet_account_interface_functions[] are part of account contracts. Where required, you must include these functions within your account contract. The logic of these functions can be mostly arbitrary, with a few limitations. For information on these limitations, see xref:#limitations_of_validation[].

[#starknet_account_interface_functions]
.Starknet account interface functions
[cols="1,3a"]
|===
| Function name | When required

| `+__validate__+` | Always required
| `+__execute__+` | Always required. The signatures of `+__validate__+` and `+__execute__+` must be identical.

[WARNING]
====
At the moment of writing (Starknet 0.13.2), two critical validations must happen in `+__execute__+`, and their absence can lead to draining of the account's funds:

(1) `assert!(get_caller_address().is_zero())`

> This asserts that the account's `execute` is not called from another contract, thus skipping validations (in later versions we may disallow calling `__execute__` from another contract at the protocol level)

Reviewer: ```suggestion
This asserts that the account's `+__execute__+` is not called from another contract, thus skipping `+__validate__+` (in future Starknet versions, we may disallow calling `__execute__` from another contract at the protocol level)
```

(2) `assert!(get_tx_info().unbox().version.into() >= 1_u32)`

> This asserts that the transaction's version is at least 1, preventing your account from accepting v0 transactions. It's critical to explicitly disallow the deprecated v0 transaction type, as v0 transactions assume that the signature verification happens in `+__execute__+`, and are thus skipping `+__validate__+` entirely.

Reviewer: ```suggestion
This asserts that the transaction's version is at least 1, preventing the account from accepting `INVOKE` v0 transactions. It is critical to explicitly disallow the deprecated v0 transaction type, as v0 transactions assume that the signature verification happens in `+__execute__+` and thus skip `+__validate__+` entirely.
```
